tests/test_rewrite.py

```diff
@@ -172,3 +173,41 @@ def test_curry_functor():
 
     rewriter = Rewriter([CurryRewriteRule()])
     assert rewriter(diagram).normal_form() == expected
+
+
+def test_unknown_words_rewrite_rule():
+    sentence = 'I see the unknown'
+    parser = BobcatParser()
+    diagram = parser.sentence2diagram(sentence).normal_form()
+
+    unknown_words = ['unknown']
+    rule = UnknownWordsRewriteRule(unknown_words=unknown_words)
+
+    rewriter = Rewriter([rule])
+    rewritten_diagram = rewriter(diagram)
+
+    expected_diagram = Diagram(dom=Ty(), cod=Ty('s'),
+            boxes=[Word('I', Ty('n')),
+                   Word('see', Ty(Ob('n', z=1), 's', Ob('n', z=-1))),
+                   Cup(Ty('n'), Ty(Ob('n', z=1))),
+                   Word('the', Ty('n', Ob('n', z=-1))),
+                   Cup(Ty(Ob('n', z=-1)), Ty('n')), Word('UNK', Ty('n')),
+                   Cup(Ty(Ob('n', z=-1)), Ty('n'))],
+            offsets=[0, 1, 0, 2, 1, 2, 1])
+
+    assert rewritten_diagram == expected_diagram
+
+def test_handle_unknown_words():
+    train_sentence = 'I love unknown unknown'
+    test_sentence = 'love unknown is what I love'
+    parser = BobcatParser()
+    train_diagrams = parser.sentence2diagram(train_sentence).normal_form()
+    test_diagrams = parser.sentence2diagram(test_sentence).normal_form()
+
+    handle_unknown_words = HandleUnknownWords(min_freq=2)
+    processed_train_diagrams = handle_unknown_words(train_diagrams, train=True, strings=None)
+    processed_test_diagrams = handle_unknown_words(test_diagrams, train=False, strings=None)
+    
+    assert handle_unknown_words.unknown_words == set(['I', 'love'])
+    assert processed_train_diagrams.boxes[0].name == 'UNK' and processed_train_diagrams.boxes[1].name == 'UNK'
+    assert processed_test_diagrams.boxes[0].name == 'UNK' and processed_test_diagrams.boxes[10].name == 'unknown'
```

Comment:
These tests are a bit funky. Any way to improve them?

Reply:
@Thommy257 Instead of an index look up, I could iterate over the list of diagrams and assert if every word box contains the token as expected. Either UNK or the word itself. Would that fit?